pallets/pooled-staking/src/lib.rs

```diff
@@ -0,0 +1,484 @@
+        >;
+
+        /// Account holding Currency of all delegators.
+        type StakingAccount: Get<Self::AccountId>;
+
+        /// When creating the first Shares for a candidate the supply can be arbitrary.
+        /// Picking a value too low will make an higher supply, which means each share will get
+        /// less rewards, and rewards calculations will have more impactful rounding errors.
+        /// Picking a value too high is a barrier of entry for staking.
+        type InitialManualClaimShareValue: Get<Self::Balance>;
+        /// When creating the first Shares for a candidate the supply can arbitrary.
+        /// Picking a value too high is a barrier of entry for staking, which will increase overtime
+        /// as the value of each share will increase due to auto compounding.
+        type InitialAutoCompoundingShareValue: Get<Self::Balance>;
+
+        /// Should be 1 to avoid rounding. (configurable for tests)
+        type InitialJoiningShareValue: Get<Self::Balance>;
+        /// Should be 1 to avoid rounding. (configurable for tests)
+        type InitialLeavingShareValue: Get<Self::Balance>;
+
+        /// Minimum amount of stake a Candidate must delegate (stake) towards itself. Not reaching
+        /// this minimum prevents from being elected.
+        type MinimumSelfDelegation: Get<Self::Balance>;
+        /// Part of the rewards that will be sent exclusively to the collator.
+        type RewardsCollatorCommission: Get<Perbill>;
+
+        /// Condition for when a joining request can be executed.
+        type JoiningRequestFilter: Contains<Self::BlockNumber>;
+        /// Condition for when a leaving request can be executed.
+        type LeavingRequestFilter: Contains<Self::BlockNumber>;
+        /// All eligible candidates are stored in a sorted list that is modified each time
+        /// delegations changes. It is safer to bound this list, in which case eligible candidate
+        /// could fall out of this list if they have less stake than the top `EligibleCandidatesBufferSize`
+        /// eligible candidates. One of this top candidates leaving will then not bring the dropped candidate
+        /// in the list. An extrinsic is available to manually bring back such dropped candidate.
+        type EligibleCandidatesBufferSize: Get<u32>;
+        /// Additional filter for candidates to be eligible.
+        type EligibleCandidatesFilter: Contains<Self::AccountId>;
+    }
+
+    /// Keeps a list of all eligible candidates, sorted by the amount of stake backing them.
+    /// This can be quickly updated using a binary search, and allow to easily take the top
+    /// `MaxCollatorSetSize`.
+    #[pallet::storage]
+    pub type SortedEligibleCandidates<T: Config> = StorageValue<
+        _,
+        BoundedVec<
+            candidate::EligibleCandidate<Candidate<T>, T::Balance>,
+            T::EligibleCandidatesBufferSize,
+        >,
+        ValueQuery,
+    >;
```

Comment:
It's surprising that this compiles when `candidate::EligibleCandidate` is a private type, I made it public in my PR #242